### DISCUSSION TITLE
How to define environment that can't be run in parallel

### DISCUSSION THREAD

**Discussion by GlenNicholls** [tox-dev/tox#2411] [Q&A] [answered]
I am slightly confused about how to control which environments can be run in parallel. For example, I have multiple environments like `py37-311` and `black`. In this scenario, when I run `tox -p`, I would like all environments to run but I don't want `black` to be run in parallel because it modifies code and thus may lead to confusing failures. I'm aware that I can use `depends` such that all envs depend on black, but this requires I run `tox -e black, py36` which is not ideal because py36 doesn't depend on black when not running in parallel. Instead, I'm trying to find a way to configure my environments so they are run either before or after envs that can be run in parallel.

If I remove `black` from `envlist` then it doesn't show up with `tox -l` but instead `tox -a`. The part I dislike about this is contributors would need to read the build plan to see there is a hidden environment that will run in CI and this is why I'm hoping to find a way to just tell tox that a specific env is not parallelizable. This way it will run on it's own during parallel runs but is also run without calling tox twice.

Any ideas?

**Comment by jugmac00** [comment, marked as answer]
You can actually use `depends` which is more a `runs_after` than really a `depends` :-)

I just verified this with the following tox configuration:
```
[tox]
envlist =
    py36,
    py37,
    py38,
    py39,
    py310,
    lint,

[testenv]
extras = test
commands = pytest {posargs}

[testenv:lint]
depends = py36,py37,py38,py39,py310
deps = pre-commit
commands = pre-commit run --all-files --show-diff-on-failure
```

And actually all envs ran in parallel, only then pre-commit, and I got the following output:
```
❯ tox -p
GLOB sdist-make: /home/jugmac00/Projects/flask-reuploaded/setup.py
✔ OK py36 in 15.53 seconds
✔ OK py310 in 20.509 seconds
✔ OK py37 in 20.538 seconds
✔ OK py38 in 20.613 seconds
✔ OK py39 in 20.709 seconds
✔ OK lint in 27.793 seconds
_________________________________________________________________________________________________________ summary __________________________________________________________________________________________________________
  py36: commands succeeded
  py37: commands succeeded
  py38: commands succeeded
  py39: commands succeeded
  py310: commands succeeded
  lint: commands succeeded
  congratulations :)
```

I used `pre-commit` instead of `black` but that does not matter.

P.S.: You still can run `black`/`pre-commit` or any test env independently.

**Reply by GlenNicholls** [reply]
I see, I suppose running it last simplifies the depends and doesn't clutter everything else up. I'll go with this for now, thank you!
